File: src/main/bridge/FrameCodec.ts
import { Buffer } from 'node:buffer';

export const BRIDGE_HEADER_SIZE = 24;
export const BRIDGE_MAX_PAYLOAD_LENGTH = 64 * 1024 * 1024;
export const BRIDGE_PROTOCOL_VERSION = 1;

export const BridgeMessageKind = {
  Handshake: 1,
  Request: 2,
  Result: 3,
  Progress: 4,
  Cancel: 5,
  Error: 6,
} as const;

export const BridgeOpcode = {
  Control: 0,
  Echo: 1,
} as const;

export const BridgeErrorCode = {
  None: 0,
  InvalidMagic: 1,
  UnsupportedVersion: 2,
  InvalidMessageKind: 3,
  InvalidFlags: 4,
  UnknownOpcode: 5,
  InvalidStatus: 6,
  InvalidRequestId: 7,
  PayloadTooLarge: 8,
  InvalidHeader: 9,
  MalformedPayload: 10,
  Cancelled: 11,
  InternalError: 12,
} as const;

export type BridgeMessageKind = typeof BridgeMessageKind[keyof typeof BridgeMessageKind];
export type BridgeOpcode = typeof BridgeOpcode[keyof typeof BridgeOpcode];
export type BridgeErrorCode = typeof BridgeErrorCode[keyof typeof BridgeErrorCode];

export interface BridgeFrame {
  kind: BridgeMessageKind;
  opcode: BridgeOpcode;
  status: BridgeErrorCode;
  requestId: number;
  payload: Uint8Array;
}

interface DecodedHeader extends Omit<BridgeFrame, 'payload'> {
  payloadLength: number;
}

const magic = Buffer.from('HFG2', 'ascii');
const validKinds = new Set<number>(Object.values(BridgeMessageKind));
const validOpcodes = new Set<number>(Object.values(BridgeOpcode));
const validErrorCodes = new Set<number>(Object.values(BridgeErrorCode).filter((value) => value !== 0));

export class BridgeProtocolError extends Error {
  readonly code: BridgeErrorCode;

  constructor(code: BridgeErrorCode, message: string) {
    super(message);
    this.name = 'BridgeProtocolError';
    this.code = code;
  }
}

export function encodeErrorMessage(message: string): Buffer {
  const bytes = Buffer.from(message, 'utf8');
  if (bytes.length === 0) {
    throw new BridgeProtocolError(BridgeErrorCode.MalformedPayload, 'Error message must not be empty');
  }
  const payload = Buffer.allocUnsafe(4 + bytes.length);
  payload.writeUInt32LE(bytes.length, 0);
  bytes.copy(payload, 4);
  return payload;
}

export function decodeErrorMessage(payload: Uint8Array): string {
  const bytes = Buffer.from(payload.buffer, payload.byteOffset, payload.byteLength);
  if (bytes.length < 4 || bytes.readUInt32LE(0) !== bytes.length - 4) {
    throw new BridgeProtocolError(BridgeErrorCode.MalformedPayload, 'Invalid error message length');
  }
  try {
    const message = new TextDecoder('utf-8', { fatal: true }).decode(bytes.subarray(4));
    if (message.length === 0) throw new Error('empty');
    return message;
  } catch {
    throw new BridgeProtocolError(BridgeErrorCode.MalformedPayload, 'Error message is not valid nonempty UTF-8');
  }
}

export function encodeFrame(frame: BridgeFrame): Buffer {
  const payload = Buffer.from(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
  validateHeader(frame.kind, 0, frame.opcode, frame.status, frame.requestId, payload.length, 0);
  if (frame.kind === BridgeMessageKind.Error) decodeErrorMessage(payload);

  const encoded = Buffer.allocUnsafe(BRIDGE_HEADER_SIZE + payload.length);
  magic.copy(encoded, 0);
  encoded.writeUInt16LE(BRIDGE_PROTOCOL_VERSION, 4);
  encoded.writeUInt8(frame.kind, 6);
  encoded.writeUInt8(0, 7);
  encoded.writeUInt16LE(frame.opcode, 8);
  encoded.writeUInt16LE(frame.status, 10);
  encoded.writeUInt32LE(frame.requestId, 12);
  encoded.writeUInt32LE(payload.length, 16);
  encoded.writeUInt32LE(0, 20);
  payload.copy(encoded, BRIDGE_HEADER_SIZE);
  return encoded;
}

export class BridgeFrameDecoder {
  readonly #header = Buffer.allocUnsafe(BRIDGE_HEADER_SIZE);
  #headerBytes = 0;
  #pending: DecodedHeader | undefined;
  #payload: Buffer | undefined;
  #payloadBytes = 0;

  push(chunk: Uint8Array): BridgeFrame[] {
    const frames: BridgeFrame[] = [];
    let offset = 0;

    while (offset < chunk.length) {
      if (!this.#pending) {
        const count = Math.min(BRIDGE_HEADER_SIZE - this.#headerBytes, chunk.length - offset);
        this.#header.set(chunk.subarray(offset, offset + count), this.#headerBytes);
        this.#headerBytes += count;
        offset += count;
        if (this.#headerBytes < BRIDGE_HEADER_SIZE) continue;

        this.#pending = decodeHeader(this.#header);
        this.#headerBytes = 0;
        if (this.#pending.payloadLength === 0) {
          frames.push(this.#complete(Buffer.alloc(0)));
          continue;
        }
        this.#payload = Buffer.allocUnsafe(this.#pending.payloadLength);
      }

      const count = Math.min(this.#pending.payloadLength - this.#payloadBytes, chunk.length - offset);
      this.#payload!.set(chunk.subarray(offset, offset + count), this.#payloadBytes);
      this.#payloadBytes += count;
      offset += count;
      if (this.#payloadBytes === this.#pending.payloadLength) frames.push(this.#complete(this.#payload!));
    }

    return frames;
  }

  complete(): void {
    if (this.#headerBytes !== 0 || this.#pending) {
      throw new BridgeProtocolError(BridgeErrorCode.MalformedPayload, 'Stream ended during a frame');
    }
  }

  #complete(payload: Buffer): BridgeFrame {
    const header = this.#pending!;
    if (header.kind === BridgeMessageKind.Error) decodeErrorMessage(payload);
    const { payloadLength: _, ...frame } = header;
    this.#pending = undefined;
    this.#payload = undefined;
    this.#payloadBytes = 0;
    return { ...frame, payload };
  }
}

function decodeHeader(header: Buffer): DecodedHeader {
  if (!header.subarray(0, 4).equals(magic)) {
    throw new BridgeProtocolError(BridgeErrorCode.InvalidMagic, 'Invalid bridge magic');
  }
  const version = header.readUInt16LE(4);
  if (version !== BRIDGE_PROTOCOL_VERSION) {
    throw new BridgeProtocolError(BridgeErrorCode.UnsupportedVersion, `Unsupported bridge version: ${version}`);
  }

  const kind = header.readUInt8(6) as BridgeMessageKind;
  const flags = header.readUInt8(7);
  const opcode = header.readUInt16LE(8) as BridgeOpcode;
  const status = header.readUInt16LE(10) as BridgeErrorCode;
  const requestId = header.readUInt32LE(12);
  const payloadLength = header.readUInt32LE(16);
  const reserved = header.readUInt32LE(20);
  validateHeader(kind, flags, opcode, status, requestId, payloadLength, reserved);
  return { kind, opcode, status, requestId, payloadLength };
}

function validateHeader(
  kind: BridgeMessageKind,
  flags: number,
  opcode: BridgeOpcode,
  status: BridgeErrorCode,
  requestId: number,
  payloadLength: number,
  reserved: number,
): void {
  if (!validKinds.has(kind)) throw new BridgeProtocolError(BridgeErrorCode.InvalidMessageKind, `Unknown message kind: ${kind}`);
  if (flags !== 0) throw new BridgeProtocolError(BridgeErrorCode.InvalidFlags, `Unsupported flags: ${flags}`);
  if (!validOpcodes.has(opcode)) throw new BridgeProtocolError(BridgeErrorCode.UnknownOpcode, `Unknown opcode: ${opcode}`);
  if (reserved !== 0) throw new BridgeProtocolError(BridgeErrorCode.InvalidHeader, 'Reserved header bytes must be zero');
  if (!Number.isInteger(requestId) || requestId < 0 || requestId > 0xffffffff) {
    throw new BridgeProtocolError(BridgeErrorCode.InvalidRequestId, `Invalid request ID: ${requestId}`);
  }
  if (!Number.isInteger(payloadLength) || payloadLength < 0 || payloadLength > BRIDGE_MAX_PAYLOAD_LENGTH) {
    throw new BridgeProtocolError(BridgeErrorCode.PayloadTooLarge, `Invalid payload length: ${payloadLength}`);
  }
  if (kind === BridgeMessageKind.Error) {
    if (!validErrorCodes.has(status)) throw new BridgeProtocolError(BridgeErrorCode.InvalidStatus, `Invalid error status: ${status}`);
  } else if (status !== BridgeErrorCode.None) {
    throw new BridgeProtocolError(BridgeErrorCode.InvalidStatus, `Non-error frame has status: ${status}`);
  }
  if (kind === BridgeMessageKind.Handshake) {
    if (opcode !== BridgeOpcode.Control || requestId !== 0) {
      throw new BridgeProtocolError(BridgeErrorCode.InvalidRequestId, 'Handshake must use control opcode and request ID zero');
    }
  } else {
    if (requestId === 0) throw new BridgeProtocolError(BridgeErrorCode.InvalidRequestId, 'Non-handshake request ID must be nonzero');
    if (kind !== BridgeMessageKind.Error && opcode === BridgeOpcode.Control) {
      throw new BridgeProtocolError(BridgeErrorCode.UnknownOpcode, 'Control opcode is only valid for handshake and error frames');
    }
  }
  if (kind === BridgeMessageKind.Cancel && payloadLength !== 0) {
    throw new BridgeProtocolError(BridgeErrorCode.MalformedPayload, 'Cancellation payload must be empty');
  }
}

File: src/Forge.Host/Bridge/FrameCodec.cs
using System.Buffers.Binary;
using System.Text;

namespace Forge.Host.Bridge;

public enum BridgeMessageKind : byte
{
    Handshake = 1,
    Request = 2,
    Result = 3,
    Progress = 4,
    Cancel = 5,
    Error = 6,
}

public enum BridgeOpcode : ushort
{
    Control = 0,
    Echo = 1,
}

public enum BridgeErrorCode : ushort
{
    None = 0,
    InvalidMagic = 1,
    UnsupportedVersion = 2,
    InvalidMessageKind = 3,
    InvalidFlags = 4,
    UnknownOpcode = 5,
    InvalidStatus = 6,
    InvalidRequestId = 7,
    PayloadTooLarge = 8,
    InvalidHeader = 9,
    MalformedPayload = 10,
    Cancelled = 11,
    InternalError = 12,
}

public sealed record BridgeFrame(
    BridgeMessageKind Kind,
    BridgeOpcode Opcode,
    BridgeErrorCode Status,
    uint RequestId,
    byte[] Payload);

public sealed class BridgeProtocolException(BridgeErrorCode code, string message) : Exception(message)
{
    public BridgeErrorCode Code { get; } = code;
}

public static class BridgeFrameCodec
{
    public const int HeaderSize = 24;
    public const int MaxPayloadLength = 64 * 1024 * 1024;
    public const ushort ProtocolVersion = 1;

    private static ReadOnlySpan<byte> Magic => "HFG2"u8;

    public static byte[] Encode(BridgeFrame frame)
    {
        ArgumentNullException.ThrowIfNull(frame.Payload);
        ValidateHeader(frame.Kind, 0, frame.Opcode, frame.Status, frame.RequestId, frame.Payload.Length, 0);
        if (frame.Kind == BridgeMessageKind.Error) BridgeErrorPayload.Decode(frame.Payload);

        var encoded = GC.AllocateUninitializedArray<byte>(HeaderSize + frame.Payload.Length);
        Magic.CopyTo(encoded);
        BinaryPrimitives.WriteUInt16LittleEndian(encoded.AsSpan(4), ProtocolVersion);
        encoded[6] = (byte)frame.Kind;
        encoded[7] = 0;
        BinaryPrimitives.WriteUInt16LittleEndian(encoded.AsSpan(8), (ushort)frame.Opcode);
        BinaryPrimitives.WriteUInt16LittleEndian(encoded.AsSpan(10), (ushort)frame.Status);
        BinaryPrimitives.WriteUInt32LittleEndian(encoded.AsSpan(12), frame.RequestId);
        BinaryPrimitives.WriteUInt32LittleEndian(encoded.AsSpan(16), (uint)frame.Payload.Length);
        BinaryPrimitives.WriteUInt32LittleEndian(encoded.AsSpan(20), 0);
        frame.Payload.CopyTo(encoded, HeaderSize);
        return encoded;
    }

    internal static DecodedHeader DecodeHeader(ReadOnlySpan<byte> header)
    {
        if (header.Length != HeaderSize)
        {
            throw new BridgeProtocolException(BridgeErrorCode.InvalidHeader, $"Header must contain {HeaderSize} bytes");
        }
        if (!header[..4].SequenceEqual(Magic))
        {
            throw new BridgeProtocolException(BridgeErrorCode.InvalidMagic, "Invalid bridge magic");
        }

        var version = BinaryPrimitives.ReadUInt16LittleEndian(header[4..]);
        if (version != ProtocolVersion)
        {
            throw new BridgeProtocolException(BridgeErrorCode.UnsupportedVersion, $"Unsupported bridge version: {version}");
        }

        var kind = (BridgeMessageKind)header[6];
        var flags = header[7];
        var opcode = (BridgeOpcode)BinaryPrimitives.ReadUInt16LittleEndian(header[8..]);
        var status = (BridgeErrorCode)BinaryPrimitives.ReadUInt16LittleEndian(header[10..]);
        var requestId = BinaryPrimitives.ReadUInt32LittleEndian(header[12..]);
        var payloadLength = BinaryPrimitives.ReadUInt32LittleEndian(header[16..]);
        var reserved = BinaryPrimitives.ReadUInt32LittleEndian(header[20..]);
        ValidateHeader(kind, flags, opcode, status, requestId, payloadLength, reserved);
        return new(kind, opcode, status, requestId, checked((int)payloadLength));
    }

    private static void ValidateHeader(
        BridgeMessageKind kind,
        byte flags,
        BridgeOpcode opcode,
        BridgeErrorCode status,
        uint requestId,
        long payloadLength,
        uint reserved)
    {
        if (!Enum.IsDefined(kind)) Throw(BridgeErrorCode.InvalidMessageKind, $"Unknown message kind: {(byte)kind}");
        if (flags != 0) Throw(BridgeErrorCode.InvalidFlags, $"Unsupported flags: {flags}");
        if (!Enum.IsDefined(opcode)) Throw(BridgeErrorCode.UnknownOpcode, $"Unknown opcode: {(ushort)opcode}");
        if (reserved != 0) Throw(BridgeErrorCode.InvalidHeader, "Reserved header bytes must be zero");
        if (payloadLength is < 0 or > MaxPayloadLength)
        {
            Throw(BridgeErrorCode.PayloadTooLarge, $"Invalid payload length: {payloadLength}");
        }

        if (kind == BridgeMessageKind.Error)
        {
            if (status == BridgeErrorCode.None || !Enum.IsDefined(status))
            {
                Throw(BridgeErrorCode.InvalidStatus, $"Invalid error status: {(ushort)status}");
            }
        }
        else if (status != BridgeErrorCode.None)
        {
            Throw(BridgeErrorCode.InvalidStatus, $"Non-error frame has status: {(ushort)status}");
        }

        if (kind == BridgeMessageKind.Handshake)
        {
            if (opcode != BridgeOpcode.Control || requestId != 0)
            {
                Throw(BridgeErrorCode.InvalidRequestId, "Handshake must use control opcode and request ID zero");
            }
        }
        else
        {
            if (requestId == 0) Throw(BridgeErrorCode.InvalidRequestId, "Non-handshake request ID must be nonzero");
            if (kind != BridgeMessageKind.Error && opcode == BridgeOpcode.Control)
            {
                Throw(BridgeErrorCode.UnknownOpcode, "Control opcode is only valid for handshake and error frames");
            }
        }

        if (kind == BridgeMessageKind.Cancel && payloadLength != 0)
        {
            Throw(BridgeErrorCode.MalformedPayload, "Cancellation payload must be empty");
        }
    }

    private static void Throw(BridgeErrorCode code, string message) => throw new BridgeProtocolException(code, message);
}

public sealed class BridgeFrameDecoder
{
    private readonly byte[] _header = GC.AllocateUninitializedArray<byte>(BridgeFrameCodec.HeaderSize);
    private int _headerBytes;
    private DecodedHeader? _pending;
    private byte[]? _payload;
    private int _payloadBytes;

    public IReadOnlyList<BridgeFrame> Push(ReadOnlySpan<byte> chunk)
    {
        var frames = new List<BridgeFrame>();

        while (!chunk.IsEmpty)
        {
            if (_pending is null)
            {
                var count = Math.Min(BridgeFrameCodec.HeaderSize - _headerBytes, chunk.Length);
                chunk[..count].CopyTo(_header.AsSpan(_headerBytes));
                _headerBytes += count;
                chunk = chunk[count..];
                if (_headerBytes < BridgeFrameCodec.HeaderSize) continue;

                _pending = BridgeFrameCodec.DecodeHeader(_header);
                _headerBytes = 0;
                if (_pending.Value.PayloadLength == 0)
                {
                    frames.Add(CompleteFrame([]));
                    continue;
                }
                _payload = GC.AllocateUninitializedArray<byte>(_pending.Value.PayloadLength);
            }

            var payloadLength = _pending.Value.PayloadLength;
            var payloadCount = Math.Min(payloadLength - _payloadBytes, chunk.Length);
            chunk[..payloadCount].CopyTo(_payload.AsSpan(_payloadBytes));
            _payloadBytes += payloadCount;
            chunk = chunk[payloadCount..];
            if (_payloadBytes == payloadLength) frames.Add(CompleteFrame(_payload!));
        }

        return frames;
    }

    public void Complete()
    {
        if (_headerBytes != 0 || _pending is not null)
        {
            throw new BridgeProtocolException(BridgeErrorCode.MalformedPayload, "Stream ended during a frame");
        }
    }

    private BridgeFrame CompleteFrame(byte[] payload)
    {
        var header = _pending!.Value;
        if (header.Kind == BridgeMessageKind.Error) BridgeErrorPayload.Decode(payload);
        var frame = new BridgeFrame(header.Kind, header.Opcode, header.Status, header.RequestId, payload);
        _pending = null;
        _payload = null;
        _payloadBytes = 0;
        return frame;
    }
}

public static class BridgeErrorPayload
{
    private static readonly UTF8Encoding StrictUtf8 = new(false, true);

    public static byte[] Encode(string message)
    {
        ArgumentNullException.ThrowIfNull(message);
        var bytes = Encoding.UTF8.GetBytes(message);
        if (bytes.Length == 0)
        {
            throw new BridgeProtocolException(BridgeErrorCode.MalformedPayload, "Error message must not be empty");
        }
        var payload = GC.AllocateUninitializedArray<byte>(4 + bytes.Length);
        BinaryPrimitives.WriteUInt32LittleEndian(payload, (uint)bytes.Length);
        bytes.CopyTo(payload, 4);
        return payload;
    }

    public static string Decode(ReadOnlySpan<byte> payload)
    {
        if (payload.Length < 4 || BinaryPrimitives.ReadUInt32LittleEndian(payload) != payload.Length - 4)
        {
            throw new BridgeProtocolException(BridgeErrorCode.MalformedPayload, "Invalid error message length");
        }
        try
        {
            var message = StrictUtf8.GetString(payload[4..]);
            return message.Length > 0
                ? message
                : throw new BridgeProtocolException(BridgeErrorCode.MalformedPayload, "Error message must not be empty");
        }
        catch (DecoderFallbackException exception)
        {
            throw new BridgeProtocolException(BridgeErrorCode.MalformedPayload, $"Error message is not valid UTF-8: {exception.Message}");
        }
    }
}

internal readonly record struct DecodedHeader(
    BridgeMessageKind Kind,
    BridgeOpcode Opcode,
    BridgeErrorCode Status,
    uint RequestId,
    int PayloadLength);
